### PR TITLE
Replace proof Verification by go-ethereum impl

### PR DESCRIPTION
**Pull request by ed255**
- Remove the custom Merkle Patricia Trie proof verification
  implementation
- Remove the custom RLP encoding/decoding implementation
- Use the Merkle Patricia Trie proof verification function from
  go-ethereum
  - This adds support to all types of non-existence proofs
- Use the RLP encoding/decoding functions from go-ethereum
- Add ~BytesHex~QuantityBytes and SliceData to abstract encoding of QUANTITY and slices of bytes in
  the StorageProof types

### CODE REVIEW

File: ethstorageproof/types.go [hunk 1]
```diff
@@ -1,25 +1,126 @@
...
+
+// UnmarshalText implements encoding.TextUnmarshaler.
+func (b *BytesHex) UnmarshalText(input []byte) error {
+	if bytes.HasPrefix(input, []byte("0x")) {
+		input = input[2:]
+	}
+	dec := make([]byte, len(input)/2)
+	if _, err := hex.Decode(dec, input); err != nil {
+		return err
+	} else {
```
Comment (p4u) [inline ethstorageproof/types.go]: No need for this else

File: ethstorageproof/types.go [hunk 2]
```diff
@@ -1,25 +1,126 @@
...
+func (b BytesHex) MarshalText() ([]byte, error) {
+	result := make([]byte, len(b)*2+2)
+	copy(result, `0x`)
+	hex.Encode(result[2:], b)
+	return result, nil
+}
+
+// UnmarshalText implements encoding.TextUnmarshaler.
+func (b *BytesHex) UnmarshalText(input []byte) error {
+	if bytes.HasPrefix(input, []byte("0x")) {
```
Comment (p4u) [inline ethstorageproof/types.go]: You can use `bytes.TrimPrefix()` instead

File: ethstorageproof/types.go [hunk 3]
```diff
@@ -1,25 +1,126 @@
...
+}
+
+// Get returns the value of the key, or nit if it's not found
+func (m *MemDB) Get(key []byte) ([]byte, error) {
+	var h common.Hash
+	copy(h[:], key)
+	value, ok := m.kvs[h]
+	if ok {
+		return value, nil
+	} else {
```
Comment (p4u) [inline ethstorageproof/types.go]: No need for this else

File: ethstorageproof/types.go [hunk 4]
```diff
@@ -1,25 +1,126 @@
...
+		return nil, nil
```
Comment (p4u) [inline ethstorageproof/types.go]: Actually I think this logic should be reversed and an error should be returned if key is not found.

```
if !ok { return nil, fmt.Errorf("...") }
return value, nil
```

File: ethstorageproof/types.go [hunk 5]
```diff
@@ -1,25 +1,126 @@
...
+	}
+}
+
+// Put sets or updates the value at key
+func (m *MemDB) Put(key []byte, value []byte) error {
+	var h common.Hash
+	copy(h[:], key)
```
Comment (p4u) [inline ethstorageproof/types.go]: You can use `common.BytesToHash()` instead

**Review by p4u — CHANGES_REQUESTED** [review]

**Comment by p4u** [comment]
It would be nice to add some extra test to this "new code". We could use the Proofs hardcoded on this test from vocdoni-node
https://github.com/vocdoni/vocdoni-node/blob/master/vochain/proof_test.go#L574

What do you think @ed255?

**Comment by ed255** [comment]
> It would be nice to add some extra test to this "new code". We could use the Proofs hardcoded on this test from vocdoni-node
> https://github.com/vocdoni/vocdoni-node/blob/master/vochain/proof_test.go#L574
> 
> What do you think @ed255?

I've added a new commit which will:
- Add proofs from https://github.com/vocdoni/vocdoni-node/blob/5b292fc7e74327f0127908636072482660b060b2/vochain/proof_test.go#L574
- Add a minime storage proof of non-existence that couldn't be verified with the previous VerifyProof implementation.

I had to edit the `value` fields from https://github.com/vocdoni/vocdoni-node/blob/5b292fc7e74327f0127908636072482660b060b2/vochain/proof_test.go#L574 because they were even-length hex strings (I added a padding 0).  How where those proofs generated?

**Comment by p4u** [comment]
RLP encoding can produce even-length hex strings, i.e 0x0 means "null" for RLP. So the Marshaling/Unmarshaling should take this into account, instead of using hex.Decode() you probably want to use some specific RLP function. Or you can just list the cases where the length can be even and modify the original hex string accordingly. 

I.e https://github.com/vocdoni/storage-proofs-eth-go/blob/master/ethstorageproof/ethstorageproof.go#L339

**Comment by p4u** [comment]
Whenever this is addresses, ping me please :)

File: ethstorageproof/ethstorageproof_test.go [hunk 6]
```diff
@@ -153,8 +154,18 @@ func TestVerify(t *testing.T) {
 	}
 
 	for i, tt := range tests {
+		var err error
+		var value []byte
+		if len(tt.value) != 0 {
+			value, err = rlp.EncodeToBytes(tt.value)
+			if err != nil {
+				panic(err)
```
Comment (p4u) [inline ethstorageproof/ethstorageproof_test.go]: We usually use t.Fatal() instead of panic()

Comment (p4u) [reply ethstorageproof/ethstorageproof_test.go]: Actually we are slowly migrating to quicktest

`import qt "github.com/frankban/quicktest"`

`qt.Assert(b, err, qt.IsNil)`

But if you don't want to change it on this PR we could do it in the future.

File: ethstorageproof/ethstorageproof_test.go [hunk 7]
```diff
@@ -172,16 +183,12 @@ func toBytes(s string) []byte {
 	return n
 }
 
-func mustRLPEncode(v interface{}) []byte {
-	value, err := rlp.EncodeToBytes(v)
-	if err != nil {
+func asQuantity(s string) []byte {
+	var q QuantityBytes
+	if err := q.UnmarshalText([]byte(s)); err != nil {
 		panic(err)
```
Comment (p4u) [inline ethstorageproof/ethstorageproof_test.go]: same

File: ethstorageproof/types.go [hunk 8]
```diff
@@ -11,45 +10,42 @@ import (
...
+type QuantityBytes []byte
 
 // MarshalText implements encoding.TextMarshaler
-func (b BytesHex) MarshalText() ([]byte, error) {
-	result := make([]byte, len(b)*2+2)
-	copy(result, `0x`)
-	hex.Encode(result[2:], b)
-	return result, nil
+func (q QuantityBytes) MarshalText() ([]byte, error) {
+	return []byte(fmt.Sprintf("0x%#x", new(big.Int).SetBytes(q))), nil
```
Comment (p4u) [inline ethstorageproof/types.go]: Can you add a comment on this expression? `0x%#x`

**Review by p4u — APPROVED** [review]

**Comment by p4u** [comment]
LGTM now, just minor comments. You can address them or let it for a future PR.

Please, "rebase and merge" yourself.

File: ethstorageproof/ethstorageproof_test.go (hunk 6 above)
Comment (ed255) [reply ethstorageproof/ethstorageproof_test.go]: I've added an issue to migrate to quicktest: https://github.com/vocdoni/storage-proofs-eth-go/issues/5

File: ethstorageproof/types.go (hunk 8 above)
Comment (ed255) [reply ethstorageproof/types.go]: I've replaced this expression to be more explicit about what's going on.  Actually there was a but, it shouldn't have bee `0x%#x` but `%#x`.  What `%#x` did was format the big.Int in hexadecimal, taking it as an integer (not as a byte slice).  I think it's more clear now, where I use `big.Int.Text(16)`.

**Pull request merged**